### PR TITLE
#479 Long text element plugin

### PR DESCRIPTION
**Pull request by CarlosNZ**
Implements #479. 

_Branched from #554 -- please review that one first._

---

Use [back-end PR #290](https://github.com/openmsupply/application-manager-server/pull/290)  to see it in action.

This is _very_ similar to the "shortText" plugin. It's almost a straight copy, but have changed `<Form.Input>` to `<Form.TextArea>`, and added a "lines" parameter to specify the height of the text area (default: 5)

Was going to add a "maxLength" parameter, but it's actually easy just to add a validation expression to check for text length if required (See example in Feature Showcase).

To test: fire up a Feature Showcase application. You'll find LongText elements on:
- Section 1, Page 3 (Need to select "Other" to enable it) -- just has default values
- Section 2, Page 4 -- lines: 8, validation expression to limit response to 100 characters

### CODE REVIEW

**Comment by nmadruga** [comment]
> Was going to add a "maxLength" parameter, but it's actually easy just to add a validation expression to check for text length if required (See example in Feature Showcase).

That called my attention when I was checking the back-end PR, I wonder if it would be nice to have something as a `maxLength` variable, just because - for now - it's such a hard thing to add a simple `validation` using REGEX.

**Comment by nmadruga** [comment]
Something weird happening when testing the Long text on Section 2 Page 4:

If I try to enter a very long response, it allows me. Even if I click out of the box - nothing changes:
Screen Shot 2021-04-12 at 1 00 37 PM

Then going to another page (previous in this case) and back to Page 4 - it's gone 😮 :
Screen Shot 2021-04-12 at 1 00 42 PM

File: src/formElementPlugins/longText/pluginConfig.json [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{
+  "isCore": true,
+  "code": "longText",
+  "displayName": "Multi-line Text Input",
+  "dynamicParameters": ["label", "description", "placeholder"],
```
Comment (nmadruga) [inline src/formElementPlugins/longText/pluginConfig.json:5]:
```suggestion
  "dynamicParameters": ["label", "description", "placeholder", "lines"],
```

File: src/formElementPlugins/longText/src/TemplateView.tsx [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import React, { useState } from 'react'
+import { Input } from 'semantic-ui-react'
+import { TemplateViewProps } from '../../types'
+
+// TODO type should be global
```
Comment (nmadruga) [inline src/formElementPlugins/longText/src/TemplateView.tsx:5]: Is this just copied from `shortText`? Not sure if that's required anymore (TODO's 👮‍♀️)

**Comment by CarlosNZ** [comment]
> Something weird happening when testing the Long text on Section 2 Page 4:
> 
> If I try to enter a very long response, it allows me. Even if I click out of the box - nothing changes:
> Screen Shot 2021-04-12 at 1 00 37 PM

It's not saving because it exceeds the maximum length. You should have seen the error state once you exceeded it, not quite sure why you didn't.
Screen Shot 2021-04-12 at 1 12 18 PM

File: src/formElementPlugins/longText/pluginConfig.json (hunk 1 above)
Comment (CarlosNZ) [reply src/formElementPlugins/longText/pluginConfig.json:5]: Do you really think "lines" needs to be dynamic? I'm struggling to think of a use case where the number of lines available in a text box would be conditional on something else.

File: src/formElementPlugins/longText/src/TemplateView.tsx (hunk 2 above)
Comment (CarlosNZ) [reply src/formElementPlugins/longText/src/TemplateView.tsx:5]: Not touching "Template View" at the moment -- please just ignore the whole file.

**Comment by CarlosNZ** [comment]
> That called my attention when I was checking the back-end PR, I wonder if it would be nice to have something as a `maxLength` variable, just because - for now - it's such a hard thing to add a simple `validation` using REGEX.

Okay, have implemented a basic `maxLength` function and updated docs as we discussed. Have also applied it to the "shortText" plugin.

File: src/formElementPlugins/longText/src/TemplateView.tsx (hunk 2 above)
Comment (nmadruga) [reply src/formElementPlugins/longText/src/TemplateView.tsx:5]: ok

**Review by nmadruga — APPROVED** [review]
Very nice. Thank you

**Pull request merged**